### ISSUE TITLE
the official 2.6 version hello-world doesn't work with standard install on Ubuntu 18

### ISSUE THREAD

**Issue by rw111** [agda/agda#4250]
Zeroth time user here. Looks like it might stay that way!
Installed Agda both via Cabal AND then via apt as well in an attempt to get stuff working first one way and then the other - time to give up...same failure result occurs whichever method used, with same error output (slight difference in search path due to .cabal versus usr/share):

```shell
Robert@laptop:\~/Code/Agda$ agda --compile hello-world.agda
Checking hello-world (/home/robert/Code/Agda/hello-world.agda).
/home/robert/Code/Agda/hello-world.agda:3,1-15
Failed to find source of module IO in any of the following
locations:
  /home/robert/Code/Agda/IO.agda
  /home/robert/Code/Agda/IO.lagda
  /home/robert/.cabal/share/x86_64-linux-ghc-8.0.2/Agda-2.6.0.1/lib/prim/IO.agda
  /home/robert/.cabal/share/x86_64-linux-ghc-8.0.2/Agda-2.6.0.1/lib/prim/IO.lagda
when scope checking the declaration
  open import IO
```

Things seem to all be in the wrong place, SINCE, if I change the hello-world program to include this line instead of the recommended line, things still don't work, but are different:

```agda
open import Agda.Builtin.IO
```

Result (I don't know why github is crossing out stuff here!!!???):

```shell
robert@laptop:~/Code/Agda$ agda --compile hello-world.agda
Checking hello-world (/home/robert/Code/Agda/hello-world.agda).
/home/robert/Code/Agda/hello-world.agda:5,8-11
Not in scope:
  run at /home/robert/Code/Agda/hello-world.agda:5,8-11
when scope checking run
robert@laptop:~/Code/Agda$ 
```

I've tried moving various library files around so that .agda and .agdai files are placed where they seem to be expected to be but then it complains about the naming conventions or something, making it impossible for a first time installer to fix by simply putting library files and Builtin files in different locations.

My conclusion is that the install process might not be working across different platforms.

**Comment by gallais** [comment]
I think the [documentation](https://agda.readthedocs.io/en/v2.6.0.1/getting-started/hello-world.html) should make it clearer that you need to have the
standard library installed and part of your `defaults` file for this compilation
command to work.

Cf. [Package system](https://agda.readthedocs.io/en/v2.6.0.1/tools/package-system.html) for a guide on how to install the stdlib & tell Agda
where to find it.

**Comment by L-TChen** [comment]
This issue also suggests that we should consider releasing something equivalent to Haskell platform so that one can quickly start their journey in Agda by conventional installation methods. 

One of difficulties of using Agda is the installation due to issues on different operation systems. The dependency on Emacs sometimes causes problems too. Package managers solve some problems of it, but not completely.

For example, Emacs inherits the environment variable `PATH` from system-wide setting instead of shell so that it complains that it cannot find `Agda-mode` if installed from scratch.  Haskell usually only sets up `PATH` for shell. It takes a moment to realise you could install a package `exec-path-from-shell` to fetch variables from shell automatically instead of always launching Emacs in a terminal emulator. Of course, this is not a problem to an Emacs expert.

Some built-in files should also be copied to a proper location like, `/use/local/lib/Agda/...` after `cabal install` or `stack install`, instead of pointing to its location in a repository.

The installation on Windows also involves a bit of trial and errors, since it depends on certain version of `icu` to work because of the outdated Haskell binding package `text-icu`.

Maybe it is a good idea to create a ticket collecting things to be done first.

**Comment by ice1000** [comment]
@L-TChen have you looked at #4244? I was thinking about the same thing, and I was trying to make Agda easy to install -- say, can be installed in a self-contained and relatively small (smaller than a complete Haskell toolchain) binary distribution. We ca  then port the package manager and things to the installer, so it's more like a **normal** programming language.
However, I'm not very confident on the quality of the installation script I wrote, and I don't know where should I post the fact that there's now an experimental way to install Agda without having to compile it, and finding interested people to help testing & fixing issues with it either (although there's still a dependency on Emacs, but you have the choice to setup only the compiler, so you can build lagda documents & blogs easily). There are also help-wanted issues too, like @lwoo1999 told me there's a linking issue on some Ubuntu server, which I can't reproduce, and I don't know how to fix it either.

**Comment by ice1000** [comment]
Also, speaking of Windows, I've tested the binary distribution I created and it works like a charm. Although I don't have a chance to test Emacs.

**Comment by rw111** [comment]
Any project with such a small community is going to have rough edges. Everybody understands that I hope. But just so you know Gallais' reply was sufficient for me to proceed (resolve this particular issue). Thanks. The quickest way to have enabled me (and therefore others) to hack it alone would have been some advice and hyperlinking between the two pages Gallais mentions, as I had no idea that I had to do more than just install stdlib (instead I was in trouble trying to figure out if perhaps apt-get install of the standard library was maybe incompatible with my Cabal install of Agda or something weird, and thus was about to give up on the whole thing - especially as there are other more cosmetic install problems in parallel with this issue, not relevant here, which made it harder for me to isolate the issue) - in the end I thought I had encountered an installation bug requiring ad hoc fixing. A third page, the install instructions page was also missing a hyperlink (or advice) to the library configuration issue: https://agda.readthedocs.io/en/v2.6.0.1/getting-started/installation.html  So two edits of the documentation would have been sufficient to make unaided install possible, even if still awkward.

**Comment by ice1000** [comment]
You can contribute to the doc. It's under the doc directory of the root dir of this project.

**Comment by nad** [comment]
We could try to make the solution to the problem a bit easier to discover by changing the error message. I suggest that we add some information to the "Failed to find source" error message when the following conditions hold:
* No options related to search paths have been given (this includes at least `-i`, `-l`, `--library-file`, `--no-libraries`, and `--no-default-libraries`). This should be easy to check now (#3567).
* The current module belongs to a project without a `.agda-lib` file. This might currently be a bit harder to check.
* Either the line `standard-library` is not included in the `defaults` file, or `standard-library.agda-lib` is not listed in the `libraries` file. This might also currently be a bit harder to check.

**Comment by samkuhn** [comment]
Just to chime in I've been trying to install Agda on OSX for the last two hours and it's pitchfork-in-the-face at every turn.. but then again I've built chrome from source so nothing phases me any more  🤣

**Comment by samkuhn** [comment]
> brew install agda

From the docs.. This is the worst thing you can do incidentally.

**Comment by UlfNorell** [comment]
Can you share your troubles? Are you trying to install the development version following https://agda.readthedocs.io/en/latest/getting-started/installation.html#installation-of-the-development-version ? Or 2.6.0.1 from Hackage?

I have not tried the brew recipe myself, so I can speak to that, but building from source should be possible without too much trouble.

**Comment by UlfNorell** [comment]
ping @mietek who is maintaining the brew recipe

**Comment by samkuhn** [comment]
Yes sorry not the most helpful comments 😉Will try and do a little writeup of my experience in the next day or so and post as a separate issue. There's a few paths to get brew / ghc / cabal / emacs / agda installed and it seems that I picked all the wrong-uns.

**Comment by mietek** [comment]
@samkuhn: Do let me know what seems to be the problem. The easiest way to talk is in the #agda IRC channel on Libera, [via IRCCloud](https://www.irccloud.com/invite?channel=%23agda&hostname=irc.libera.chat&port=6697&ssl=1) or otherwise.

**Comment by samkuhn** [comment]
Hi @mietek - sure happy to, a little busy right at the moment with a few deadlines but will retrace my steps and post back soon. It may be that I was just tired and missed something obvious at the start. Or perhaps just an old version of Homebrew. Also not having Haskell set up here beforehand might have made things a little more corner-case.

**Comment by samkuhn** [comment]
Hey @mietek, @UlfNorell just to follow up on this - I removed everything and attempted to retrace my steps. While it wasn't perfectly smooth sailing it did install successfully this time (whoop!).

I've uploaded some notes here: #4497 if it's useful at all. I'm not entirely sure what happened last time although I know something made me stop trying Homebrew and begin installing Haskell manually first and attempting to install Agda via cabal instead..

Super exited to be getting stuck into some Agda coding and apologies for hijacking this thread.

**Comment by capn-freako** [comment]
> Also, speaking of Windows, I've tested the binary distribution I created and it works like a charm. Although I don't have a chance to test Emacs.

@ice1000 , can you provide some brief instructions for using the two *.MSI files here: http://homepage.divms.uiowa.edu/~astump/agda/
?

Thanks!

**Comment by capn-freako** [comment]
I'm really confused, because I seem to have the standard library installed.
That is, I see lots of `*.agda` and `*.agdai` files in:

- `C:\Program Files\Haskell\x86_64-windows-ghc-8.6.5\Agda-2.6.0.1\lib\prim\Agda\Builtin\`, and
- `C:\Program Files\Haskell\x86_64-windows-ghc-8.6.5\Agda-2.6.0.1\lib\prim\Agda\Primitive\`,

including the infamous `IO.agda`.

But, I can't find a `standard-library.agda-lib`file anywhere.

Did I manage to create a "half installation" somehow?
I used the `http://homepage.divms.uiowa.edu/~astump/agda/Agda2.6.0.1.v1.msi` provided by @ice1000 . But, I did *not* install the `http://homepage.divms.uiowa.edu/~astump/agda/AgdaBundle_2.5.4.2.v4.msi` file first.
Was I supposed to?

Any thoughts/help would be most appreciated.

Has there emerged, yet, a clearly superior method for getting quickly up and running w/ *Agda* on Windows 10 systems?

Thanks!
Can't wait to start playing.
:)

P.S. - Oh, I forgot to mention: when I attempt to compile the `hello-world.agda` file, I get:

```
Failed to find source of module IO in any of the following
locations:
  C:\Users\capnf\Documents\Agda\IO.agda
  C:\Users\capnf\Documents\Agda\IO.lagda
  C:\Program Files\Haskell\x86_64-windows-ghc-8.6.5\Agda-2.6.0.1\lib\prim\IO.agda
  C:\Program Files\Haskell\x86_64-windows-ghc-8.6.5\Agda-2.6.0.1\lib\prim\IO.lagda
when scope checking the declaration
  open import IO
```

**Comment by gallais** [comment]
> That is, I see lots of *.agda and *.agdai files in:

That is just the built-in modules. I don't think Aaron uses the standard library in his lectures so it is probably not shipped with the bundle and you need to install it separately. You can find the installation instructions on the stdlib repo (be careful to pick a version compatible with your Agda version): https://github.com/agda/agda-stdlib/#installation-instructions

**Comment by robwithhair** [comment]
I just installed Agda on two Macs.  First with stack and second with Cabal.  I got the same error with both machines when running the hello world from the docs and it seems to be to do with the run function.  This is after installing the standard library version 1.5
![image](https://user-images.githubusercontent.com/6692245/106636129-831b8880-6579-11eb-9a2f-ae2ab85395a6.png)
![image](https://user-images.githubusercontent.com/6692245/106636149-8adb2d00-6579-11eb-859c-2a0ed9c1f478.png)
```
Agda version 2.6.1.2
agda-stdlib-1.5
```
I'd really appreciate any input as to what I'm doing wrong?  I was hoping to get using Agda as a proof assistant today. :-(

**Comment by gallais** [comment]
We've made `putStrLn` more level polymorphic in stdlib v1.5. It should be enough to
add a type annotation to `main`. Something like:

```agda
module hello-world where

open import IO
open import Level

main = run {0ℓ} (putStrLn "Hello, World!")
```

**Comment by jespercockx** [comment]
The PLFA book has a different "hello world" program that doesn't require `IO` or the standard library:
```agda
data Greeting : Set where
  hello : Greeting

greet : Greeting
greet = hello
```
Since being able to load this file is sufficient to start using Agda as a proof assistant, perhaps we should add it to the user manual before the current example, and mention that the current "hello world" is only required for testing if you have the standard library installed and are able to compile Agda programs.

**Comment by MatthewDaggitt** [comment]
I strongly agree with @jespercockx. I imagine that most new users are not immediately looking to do IO in Agda. I myself have _never_ performed any IO in Agda in my 4 years of using it!

**Comment by andreasabel** [comment]
> The PLFA book has a different "hello world" program that doesn't require `IO` or the standard library:
> 
> ```agda
> data Greeting : Set where
>   hello : Greeting
> 
> greet : Greeting
> greet = hello
> ```

Truly in the spirit of Agda.  A program isn't something that runs, it is something that type-checks! :-D

**Comment by gallais** [comment]
> I imagine that most new users are not immediately looking to do IO in Agda.

On the other hand, users that do want to do some IO (e.g. people
who do "advent of code"-like challenges for fun) and are looking
into expanding their horizons by dipping their toes in a dependently
typed programming language may be put off by Agda's lack of basic
IO examples.

**Comment by MatthewDaggitt** [comment]
> On the other hand, users that do want to do some IO (e.g. people
who do "advent of code"-like challenges for fun) and are looking
into expanding their horizons by dipping their toes in a dependently
typed programming language may be put off by Agda's lack of basic
IO examples.

Agreed. As far as I understood, @jespercockx wasn't advocating for replacing the current IO example, only inserting the non-IO example in front of it.

**Comment by fedvasu** [comment]
I always end up with the following error,

agda: /usr/share/libghc-agda-dev/MAlonzo/src: getDirectoryContents:openDirStream: does not exist (No such file or directory)

I have tried installing through, ubuntu package manager, cabal and stack.

None of them work with any std-lib, either package or directly installing through github.

None of the solutions work for any kind of IO. 

I followed all of the above comments carefully, none of those actually solve issue related to this. Not using IO is not a solution. I read the install and hello world documents many times over, I just can't get it to compile a hello world program.

**Comment by gallais** [comment]
> agda: /usr/share/libghc-agda-dev/MAlonzo/src: getDirectoryContents:openDirStream: does not exist (No such file or directory)

This suggests it's the version installed by the package manager that gets used.
Have you uninstalled it before trying cabal or stack? If you type `which agda`
in a terminal it should tell you which one it's picking up. This way you can
make sure you're not repeatedly trying the same one despite different installs.
